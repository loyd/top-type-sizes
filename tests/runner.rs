use top_type_sizes::*;

macro_rules! test_sample {
    ($name:ident) => {
        #[test]
        fn $name() {
            static SAMPLE: &'static str =
                include_str!(concat!("samples/", stringify!($name), ".txt"));

            let refined_sample = reader::read(SAMPLE.as_bytes()).unwrap();
            let types = parser::parse(&refined_sample).unwrap();
            insta::assert_debug_snapshot!(format!("{}_types", stringify!($name)), types);

            // TODO: check the transformer.

            let output = formatter::format(types, &Default::default());
            println!("{output}");
            // TODO: make snapshots of output.
        }
    };
}

test_sample!(command);
test_sample!(control_flow);
test_sample!(alignment_enum);
test_sample!(tokio_udp);
test_sample!(several_types);
test_sample!(timex);
test_sample!(async_fn);
test_sample!(compiler_messages);

// TODO: add samples from rustc tests.
