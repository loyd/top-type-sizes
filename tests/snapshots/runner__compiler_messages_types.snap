---
source: tests/runner.rs
expression: types
---
[
    Type {
        name: "{async fn body of run()}",
        size: 7880,
        align: 8,
        kind: Enum(
            EnumType {
                discriminant_size: Some(
                    1,
                ),
                variants: [
                    EnumVariant {
                        name: "Unresumed",
                        size: 592,
                        items: [
                            Field(
                                Field {
                                    kind: Upvar,
                                    name: "self",
                                    size: 568,
                                    align: Some(
                                        8,
                                    ),
                                    offset: Some(
                                        0,
                                    ),
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: Upvar,
                                    name: "drain",
                                    size: 24,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                        ],
                    },
                    EnumVariant {
                        name: "Suspend0",
                        size: 7879,
                        items: [
                            Field(
                                Field {
                                    kind: Upvar,
                                    name: "self",
                                    size: 568,
                                    align: Some(
                                        8,
                                    ),
                                    offset: Some(
                                        0,
                                    ),
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: Upvar,
                                    name: "drain",
                                    size: 24,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Field(
                                Field {
                                    kind: GeneratorLocal,
                                    name: "self",
                                    size: 568,
                                    align: None,
                                    offset: None,
                                    local_type: None,
                                },
                            ),
                            Padding(
                                1,
                            ),
                            Field(
                                Field {
                                    kind: GeneratorLocal,
                                    name: "coroutine_field3",
                                    size: 1,
                                    align: Some(
                                        1,
                                    ),
                                    offset: None,
                                    local_type: Some(
                                        "bool",
                                    ),
                                },
                            ),
                            Padding(
                                6,
                            ),
                            Field(
                                Field {
                                    kind: GeneratorLocal,
                                    name: "__awaitee",
                                    size: 6712,
                                    align: Some(
                                        8,
                                    ),
                                    offset: None,
                                    local_type: Some(
                                        "{async fn body of run()}",
                                    ),
                                },
                            ),
                        ],
                    },
                ],
            },
        ),
        end_padding: None,
    },
]
